### ISSUE TITLE
Issue inserting into blockdata2

### ISSUE THREAD

**Issue by jonathanjm** [ethgasstation/ethgasstation-backend#20]
Hi,
I've been having some issues trying to run the backend. I'm new to Python and have been having problems debugging this one:

`File "/home/jonathan/ethgasstation-backend/egs/main.py", line 240, in update_dataframes
  write_to_sql(alltx, block_sumdf, mined_blockdf, block)
File "/home/jonathan/ethgasstation-backend/egs/main.py", line 76, in write_to_sql
  block_sumdf.to_sql(con=engine, name='blockdata2', if_exists='append', index=False)
File "/home/jonathan/.local/lib/python3.5/site-packages/pandas/core/generic.py", line 1534, in to_sql
  chunksize=chunksize, dtype=dtype)
File "/home/jonathan/.local/lib/python3.5/site-packages/pandas/io/sql.py", line 473, in to_sql
  chunksize=chunksize, dtype=dtype)
File "/home/jonathan/.local/lib/python3.5/site-packages/pandas/io/sql.py", line 1156, in to_sql
  table.insert(chunksize)
File "/home/jonathan/.local/lib/python3.5/site-packages/pandas/io/sql.py", line 670, in insert
  self._execute_insert(conn, keys, chunk_iter)
File "/home/jonathan/.local/lib/python3.5/site-packages/pandas/io/sql.py", line 645, in _execute_insert
  conn.execute(self.insert_statement(), data)
File "/home/jonathan/.local/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 945, in execute
  return meth(self, multiparams, params)
File "/home/jonathan/.local/lib/python3.5/site-packages/sqlalchemy/sql/elements.py", line 263, in _execute_on_connection
  return connection._execute_clauseelement(self, multiparams, params)
File "/home/jonathan/.local/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1053, in _execute_clauseelement
  compiled_sql, distilled_params
File "/home/jonathan/.local/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1189, in _execute_context
  context)
File "/home/jonathan/.local/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1402, in _handle_dbapi_exception
  exc_info
File "/home/jonathan/.local/lib/python3.5/site-packages/sqlalchemy/util/compat.py", line 203, in raise_from_cause
  reraise(type(exception), exception, tb=exc_tb, cause=cause)
File "/home/jonathan/.local/lib/python3.5/site-packages/sqlalchemy/util/compat.py", line 186, in reraise
  raise value.with_traceback(tb)
File "/home/jonathan/.local/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1182, in _execute_context
  context)
File "/home/jonathan/.local/lib/python3.5/site-packages/sqlalchemy/engine/default.py", line 470, in do_execute
  cursor.execute(statement, parameters)
File "/home/jonathan/.local/lib/python3.5/site-packages/mysql/connector/cursor.py", line 547, in execute
  stmt, self._process_params_dict(params))
File "/home/jonathan/.local/lib/python3.5/site-packages/mysql/connector/cursor.py", line 414, in _process_params_dict
  "Failed processing pyformat-parameters; %s" % err)
sqlalchemy.exc.ProgrammingError: (mysql.connector.errors.ProgrammingError) Failed processing pyformat-parameters; Python 'hexbytes' cannot be converted to a MySQL type [SQL: 'INSERT INTO blockdata2 (miner, blockhash, includedblock, block_number, mingasprice, time_mined, gaslimit, blockfee, numtx, gasused, speed, main) VALUES (%(miner)s, %(blockhash)s, %(includedblock)s, %(block_number)s, %(mingasprice)s, %(time_mined)s, %(gaslimit)s, %(blockfee)s, %(numtx)s, %(gasused)s, %(speed)s, %(main)s)'] [parameters: {'miner': '0xa0a0f44f1b7a9dbf2e7a7b3622a497302d30d9944527f17c9ea5edd1d71d9400', 'blockhash': HexBytes('0x2b2c08a4cf048b2f0cd9ad463a6826c84ce20d611e8c6a53e4f53494aec5ab38'), 'numtx': None, 'block_number': 313269, 'mingasprice': None, 'time_mined': 0, 'gaslimit': 14995606, 'blockfee': 0.0, 'includedblock': None, 'main': 1, 'speed': 0.0, 'gasused': 0}]
`

any ideas why?

Thank you

**Comment by ethgasstation** [comment]
Hi- yes- we just pushed an upgrade to the backend which supports web3.py version 4 (which uses hexbytes to store transaction data).  If you update the backend, it should work now.

**Comment by jonathanjm** [comment]
Hi,

Thanks for the update!

I updated but still getting the same error.
I didn't include this in my last post:
`[ERROR] Traceback (most recent call last):
  File "/home/jonathan/.local/lib/python3.5/site-packages/mysql/connector/conversion.py", line 184, in to_mysql
    return getattr(self, "_{0}_to_mysql".format(type_name))(value)
AttributeError: 'MySQLConverter' object has no attribute '_hexbytes_to_mysql' `

Do I need to update the mysql-connector-python?

Thanks

**Comment by ethgasstation** [comment]
Wow- good timing- I just fixed this now after your note- can you update your repo again and see if it is fixed now?  You don't need to update mysql-connector.

**Comment by jonathanjm** [comment]
:) thanks. I'll give it a try now

**Comment by ethgasstation** [comment]
I just added one more fix if you want to run with the —generate-report option

**Comment by jonathanjm** [comment]
It fixed the issue above but now I'm getting this error from one of the code changes you made recently:
[ERROR] Traceback (most recent call last):
  File "/home/jonathan/ethgasstation-backend/egs/main.py", line 235, in update_dataframes
    report = SummaryReport(last1500t, last1500b, block)
  File "/home/jonathan/ethgasstation-backend/egs/report_generator.py", line 50, in __init__
    self.post['cheapestTxID'] = (self.tx_df.loc[(self.tx_df['minedGasPrice']==self.post['cheapestTx']) & (self.tx_df['gas_offered'] == 21000)].index[0]).lower()
  File "/home/jonathan/.local/lib/python3.5/site-packages/pandas/core/indexes/base.py", line 2069, in __getitem__
    return getitem(key)
IndexError: index 0 is out of bounds for axis 0 with size 0

**Comment by ethgasstation** [comment]
Is this just a one- time error or does it happen each time? It means there were no transactions with 21000 gasused (ie transfers) in your dataframe when the report was run.  It shouldn’t crash the script

**Comment by jonathanjm** [comment]
Yea, it keeps happening and it's not generating the .json files. I commented out that line for now to be able to get the files and noticed that `self.post['avgTxFee'] = self.tx_df.loc[self.tx_df['gas_offered']==21000, 'minedGasPrice'].median()` is returning NAN.

I tried to debug this and also noticed that `web3.txpool.content` is returning an empty array, I think this is a problem with my node and not your code.
